### ISSUE TITLE
{"port":80} directive does not pick up the right port when more then one port is published and containers are added and/or removed

### ISSUE THREAD

**Issue by twobombs** [ehazlett/interlock#79]
docker run -p 80:80 --name interlock --restart=always --hostname interlock -e INTERLOCK_DATA='{"port":80}' -e HAPROXY_PROXY_BACKEND_OVERRIDE_ADDRESS=192.168.99.100 -e HAPROXY_CLIENT_TIMEOUT=2000000 -e HAPROXY_SERVER_TIMEOUT=2000000 ehazlett/interlock -s tcp://192.168.99.100:2375 --plugin haproxy start

does not always pick up port 80 of a container 

log snippet:

works and publishes correct: ( chance ~ 1 in 10 )

"2016-02-29T11:34:37.669957771Z time="2016-02-29T11:34:37Z" level="info" msg="[haproxy] [<URL withheld>]: upstream=<IP>:30001 container=<URL withheld>"

but when a container is added or removed the following pops up:

2016-02-29T12:15:05.495103290Z time="2016-02-29T12:15:05Z" level="info" msg="[haproxy] <URL withheld>: upstream=<IP>:30003 container=<URL withheld>" 

port 30001 is port 80, 30003 is MySQL and 30002 is ssh

We see similar behaviour when ports are dynamically allocated and with fewer ports publised ( eg: 80 & ssh ) with only one port this does work.

**Comment by ehazlett** [comment]
This looks like it is working as expected:

The following gets port "80" exposed as it's the only port.

`docker run --rm -ti -p 80:80 --hostname foo.local alpine ash`

```
interlock:ehazlett interlock> ./cmd/interlock/interlock -D run -c ~/Sync/interlock.conf.haproxy                          
INFO[0000] interlock 1.0.0 (0f87cb0)                    
DEBU[0000] docker client: url=unix:///var/run/docker.sock 
DEBU[0000] loading extension: name=haproxy configpath=/tmp/haproxy.cfg 
DEBU[0000] starting event handling                      
DEBU[0000] checking to reload                           
DEBU[0000] reloading                                    
DEBU[0000] updating load balancers                      
DEBU[0000] generating proxy config                       ext=haproxy
INFO[0000] configuration updated                         ext=haproxy
DEBU[0000] dropping SYN packets to trigger client re-send  ext=haproxy
DEBU[0000] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0000] error signaling clients to resend; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0000] resuming SYN packets                          ext=haproxy
DEBU[0000] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0000] error signaling clients to resume; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0000] reload duration: 14.13ms                     
DEBU[0010] event received: type=create id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0010] event received: type=attach id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0010] event received: type= id=                    
DEBU[0010] event received: type=start id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0010] inspecting container: id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0010] checking container labels: id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0010] checking container ports: id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0010] container is monitored; triggering reload: id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0010] triggering reload                            
DEBU[0010] event received: type=resize id=701d1aedb4fc3777bc4e95815918ce0ec2b86ad867fe3f34ab5e033bcb946943 
DEBU[0011] reaping key: reload                          
DEBU[0011] triggering reload from cache                 
DEBU[0011] checking to reload                           
DEBU[0011] reloading                                    
DEBU[0011] updating load balancers                      
DEBU[0011] generating proxy config                       ext=haproxy
INFO[0011] foo.local: upstream=172.17.0.1:80 container=determined_wozniak  ext=haproxy
DEBU[0011] alias domains: []                             ext=haproxy
DEBU[0011] adding host name=foo_local domain=foo.local   ext=haproxy
INFO[0011] configuration updated                         ext=haproxy
```

If more than one port is published I can specify which port to use:

`docker run --rm -ti -p 80:80 -p 3000:3000 --label interlock.port=80 --hostname foo.local  alpine ash`

You can see the upstream is `172.17.0.1:80`

```
INFO[0000] interlock 1.0.0 (0f87cb0)                    
DEBU[0000] docker client: url=unix:///var/run/docker.sock 
DEBU[0000] loading extension: name=haproxy configpath=/tmp/haproxy.cfg 
DEBU[0000] starting event handling                      
DEBU[0000] checking to reload                           
DEBU[0000] reloading                                    
DEBU[0000] updating load balancers                      
DEBU[0000] generating proxy config                       ext=haproxy
INFO[0000] configuration updated                         ext=haproxy
DEBU[0000] dropping SYN packets to trigger client re-send  ext=haproxy
DEBU[0000] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0000] error signaling clients to resend; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0000] resuming SYN packets                          ext=haproxy
DEBU[0000] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0000] error signaling clients to resume; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0000] reload duration: 12.87ms                     
DEBU[0007] event received: type=create id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0007] event received: type=attach id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0007] event received: type= id=                    
DEBU[0007] event received: type=start id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0007] inspecting container: id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0007] checking container labels: id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0007] checking container ports: id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0007] container is monitored; triggering reload: id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0007] triggering reload                            
DEBU[0007] event received: type=resize id=07223ce67654856216f540cfcf5d1b08af79c9766f61671578d554d7e796cd5c 
DEBU[0008] reaping key: reload                          
DEBU[0008] triggering reload from cache                 
DEBU[0008] checking to reload                           
DEBU[0008] reloading                                    
DEBU[0008] updating load balancers                      
DEBU[0008] generating proxy config                       ext=haproxy
DEBU[0008] foo.local: found specified port 80 exposed as 80  ext=haproxy
INFO[0008] foo.local: upstream=172.17.0.1:80 container=admiring_ptolemy  ext=haproxy
DEBU[0008] alias domains: []                             ext=haproxy
DEBU[0008] adding host name=foo_local domain=foo.local   ext=haproxy
INFO[0008] configuration updated                         ext=haproxy
DEBU[0008] dropping SYN packets to trigger client re-send  ext=haproxy
DEBU[0008] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0008] error signaling clients to resend; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0008] resuming SYN packets                          ext=haproxy
DEBU[0008] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0008] error signaling clients to resume; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0008] reload duration: 13.90ms                     
```

I can change the port specification for interlock and it adheres:

`docker run --rm -ti -p 80:80 -p 3000:3000 --label interlock.port=3000 --hostname foo.local  alpine ash`

Here I'm specifying port `3000` instead of `80`.

```
INFO[0000] interlock 1.0.0 (0f87cb0)                    
DEBU[0000] docker client: url=unix:///var/run/docker.sock 
DEBU[0000] loading extension: name=haproxy configpath=/tmp/haproxy.cfg 
DEBU[0000] starting event handling                      
DEBU[0000] checking to reload                           
DEBU[0000] reloading                                    
DEBU[0000] updating load balancers                      
DEBU[0000] generating proxy config                       ext=haproxy
INFO[0000] configuration updated                         ext=haproxy
DEBU[0000] dropping SYN packets to trigger client re-send  ext=haproxy
DEBU[0000] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0000] error signaling clients to resend; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0000] resuming SYN packets                          ext=haproxy
DEBU[0000] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0000] error signaling clients to resume; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0000] reload duration: 14.54ms                     
DEBU[0004] event received: type=create id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0004] event received: type=attach id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0004] event received: type= id=                    
DEBU[0004] event received: type=start id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0004] inspecting container: id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0004] checking container labels: id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0004] checking container ports: id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0004] container is monitored; triggering reload: id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0004] triggering reload                            
DEBU[0004] event received: type=resize id=e6cbd64c80f8e5577662ee806d6a3a321040b1230478c836dc9b58feb6b9326c 
DEBU[0005] reaping key: reload                          
DEBU[0005] triggering reload from cache                 
DEBU[0005] checking to reload                           
DEBU[0005] reloading                                    
DEBU[0005] updating load balancers                      
DEBU[0005] generating proxy config                       ext=haproxy
DEBU[0005] foo.local: found specified port 3000 exposed as 3000  ext=haproxy
INFO[0005] foo.local: upstream=172.17.0.1:3000 container=sleepy_almeida  ext=haproxy
DEBU[0005] alias domains: []                             ext=haproxy
DEBU[0005] adding host name=foo_local domain=foo.local   ext=haproxy
INFO[0005] configuration updated                         ext=haproxy
DEBU[0005] dropping SYN packets to trigger client re-send  ext=haproxy
DEBU[0005] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0005] error signaling clients to resend; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0005] resuming SYN packets                          ext=haproxy
DEBU[0005] &{/sbin/iptables [/sbin/iptables -I INPUT -p tcp --dport 80 --syn -j DROP] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}  ext=haproxy
WARN[0005] error signaling clients to resume; you will notice dropped packets: exit status 3  ext=haproxy
DEBU[0005] reload duration: 15.45ms 
```

You can see the upstream selected is using port `3000`.

**Comment by ehazlett** [comment]
Ah so the example I used is for the new version.  Let me debug the legacy version to see if I can re-create.

**Comment by twobombs** [comment]
Thank you for the response, really grateful for this. The "--label interlock.port=3000" is a new one, looks as though it's usefull for the static mapped ports.

However, mapping is mostly done on a dynamic port eg:

docker run --rm -ti -p 80 -p 8022 --hostname foo.local alpine ash 

With the 'port'=80 directive oftentimes port 8022 or port 3306 instead of port 80 is relayed through HAPROXY 

<img width="467" alt="screenshot 2016-03-03 15 00 22" src="https://cloud.githubusercontent.com/assets/12692227/13496485/b99d338e-e150-11e5-98a3-21da205ad910.png">

<img width="467" alt="screenshot 2016-03-03 15 00 22" src="https://cloud.githubusercontent.com/assets/12692227/13496492/c0178ca0-e150-11e5-85aa-9dee3853bb70.png">

**Comment by ehazlett** [comment]
Ok this is what the `{"port":80}` is supposed to fix.  Let me see if I can re-create on legacy.

As another thought, you might want to try the new version (1.0) although this does require the load balancing container be external (i.e. another nginx or haproxy container).

I will see if I can re-create this scenario.  Thanks!

**Comment by ehazlett** [comment]
This looks like it's working in legacy (0.3.3) as well:

```
time="2016-03-03T09:29:50Z" level=debug msg="[interlock] event: date=1456997390 type=resize image=alpine container=4486035da8d05138d8efea0fc65da11e1e17c305ae0c3a14511839d8990de449"
time="2016-03-03T09:29:50Z" level=info msg="[interlock] dispatching event to plugin: name=nginx version=0.1"
time="2016-03-03T09:29:50Z" level=info msg="[nginx] action=received event=4486035da8d05138d8efea0fc65da11e1e17c305ae0c3a14511839d8990de449 time=1456997390"
time="2016-03-03T09:29:50Z" level=debug msg="[nginx] foo.local: found specified port 8000 exposed as 32775"
time="2016-03-03T09:29:50Z" level=info msg="[nginx] foo.local: upstream=0.0.0.0:32775"
time="2016-03-03T09:29:50Z" level=debug msg="[nginx] writing config:  /etc/nginx/nginx.conf"
time="2016-03-03T09:29:50Z" level=debug msg="[nginx] reloading nginx process"
time="2016-03-03T09:29:50Z" level=info msg="[nginx] nginx reloaded and ready"
```

You can see in the logs "found specified port 8000...".  Can you give the version of Interlock you are using as well as the logs from Interlock when you run one of these containers having issues?

**Comment by twobombs** [comment]
<img width="883" alt="screenshot 2016-03-03 15 41 39" src="https://cloud.githubusercontent.com/assets/12692227/13497584/7459edde-e156-11e5-8f74-f2ad835b0161.png">

version obtained with the docker run command as mentioned in the first post ( pulls the default version )

Below a fail ( publishing port 30003 )
![screenshot 2016-03-03 15 44 17](https://cloud.githubusercontent.com/assets/12692227/13497913/eb8a7f4e-e157-11e5-9af6-8dd857e94e61.png)

And here good one ( publishing 30001 )
![screenshot 2016-03-03 15 44 17](https://cloud.githubusercontent.com/assets/12692227/13497917/f10cc65c-e157-11e5-9b3e-7fb16933c018.png)

**Comment by ehazlett** [comment]
Thanks. This looks like an older version of interlock. Could you pull the
latest (docker pull ehazlett/interlock:latest) and try again?

**Comment by twobombs** [comment]
We see it also failing with v0.33 : 

2016-03-04T10:49:11.721971460Z time="2016-03-04T10:49:11Z" level=info msg="[haproxy]  upstream=192.168.0.224:32768 container=

2016-03-04T10:49:40.303193956Z time="2016-03-04T10:49:40Z" level=info msg="[haproxy] upstream=192.168.0.224:32769 container=

flipping between the two published ports

**Comment by ehazlett** [comment]
Hmm.  Can you give the entire log?  I cannot re-create when I specify the port in the environment variable.  If you do not set that it will select the first published port which would cause this action.

**Comment by twobombs** [comment]
yupz, 0.33 has this fixed IMHO, all this time we ran with 0.31, the 0.33 version works(tm)

thank you very much for the information, we can now more ahead with mass-deployment ! :)

**Comment by twobombs** [comment]
update: dynamic allocated ports works
static ports do not

log:

2016-03-08T13:18:45.649614494Z time="2016-03-08T13:18:45Z" level=warning msg="***** THIS VERSION IS DEPRECATED.  PLEASE UPDATE *****" 
2016-03-08T13:18:45.649780736Z time="2016-03-08T13:18:45Z" level=warning msg="See https://github.com/ehazlett/interlock for details" 
w2016-03-08T13:18:45.650604038Z time="2016-03-08T13:18:45Z" level=info msg="interlock running version=0.3.3 (31890d8)" 
u2016-03-08T13:18:45.650821425Z time="2016-03-08T13:18:45Z" level=info msg="loading plugin name=haproxy version=0.1" 
2016-03-08T13:18:45.650838342Z time="2016-03-08T13:18:45Z" level=info msg="[interlock] dispatching event to plugin: name=haproxy version=0.1" 
2016-03-08T13:18:45.666240480Z time="2016-03-08T13:18:45Z" level=info msg="[haproxy] <URL>: upstream=192.168.0.222:30003

<img width="458" alt="screenshot 2016-03-08 14 41 58" src="https://cloud.githubusercontent.com/assets/12692227/13603229/f3c9024c-e53b-11e5-981c-226e246684f6.png">

**Comment by ehazlett** [comment]
Hmm can you give a little more logs?  As you can see in my previous log output I have both dynamic and manually specified ports and they both are working.

**Comment by ehazlett** [comment]
It also looks like those are dynamic only as in they are allocating high ports.

**Comment by twobombs** [comment]
The port is high yet static, however, as the dynamic port allocation now works i regard this issue as closed, also because v1.0 has been released ( and this version is no longer maintained ) Will look into a 1.0 migration in the coming months. Thank you very much for all your hard work.

**Comment by ehazlett** [comment]
Sounds good thank you!
